include/eosio/vm/argument_proxy.hpp

```diff
@@ -0,0 +1,120 @@
+      static_assert(std::is_trivially_copyable_v<T>, "argument_proxy requires a trivially copyable type");
+      inline constexpr argument_proxy(void* ptr) : original_ptr(ptr) {
+         if (!LegacyAlign || reinterpret_cast<std::uintptr_t>(original_ptr) % LegacyAlign != 0) {
+            copy.emplace();
+            memcpy( std::addressof(*copy), original_ptr, sizeof(T) );
+         }
+      }
+      inline constexpr argument_proxy(const argument_proxy&) = delete;
+      inline constexpr argument_proxy(argument_proxy&& other) : original_ptr(other.original_ptr), copy(other.copy) {
+         other.copy.reset();
+         other.original_ptr = nullptr;
+      }
+      inline ~argument_proxy() {
+         if constexpr (!std::is_const_v<T>)
+            if (copy)
+               memcpy( original_ptr, std::addressof(*copy), sizeof(T) );
+      }
+      constexpr operator T*() { return get(); }
+      constexpr operator const T*() const { return get(); }
+
+      constexpr T* get() { return const_cast<T*>(const_cast<const argument_proxy*>(this)->get()); }
+      constexpr const T* get() const {
+         if (copy)
+            return std::addressof(*copy);
+         else
+            return static_cast<const T*>(original_ptr);
+      }
+      constexpr T* operator->() { return get(); }
+      constexpr const T* operator->() const { return get(); }
+
+      static constexpr bool is_legacy() { return LegacyAlign != 0; }
+      using dependent_type = T;
+
+      void* original_ptr;
+      std::optional<std::remove_cv_t<T>> copy;
+   };
+
+   template <typename T, std::size_t LegacyAlign>
+   struct argument_proxy<span<T>, LegacyAlign> : span<T> {
+      static_assert(LegacyAlign % alignof(T) == 0, "Specified alignment must be at least alignment of T");
+      static_assert(std::is_trivially_copyable_v<T>, "argument_proxy requires a trivially copyable type");
+      inline constexpr bool is_aligned(void* ptr) { return reinterpret_cast<std::uintptr_t>(original_ptr) % LegacyAlign == 0; }
+      inline constexpr argument_proxy(void* ptr, uint32_t size)
+         : original_ptr(ptr),
+           copy( is_aligned(ptr) ? nullptr : new std::remove_cv_t<T>[size] ) {
+         *static_cast<span<T>*>(this) = span<T>( copy ? copy.get() : static_cast<T*>(ptr), size );
+         if (copy)
+            memcpy( copy.get(), original_ptr, this->size_bytes() );
+      }
+      inline constexpr argument_proxy(const argument_proxy&) = delete;
+      inline constexpr argument_proxy(argument_proxy&&) = default;
+      inline ~argument_proxy() {
+         if constexpr (!std::is_const_v<T>)
+            if (copy)
+               memcpy( original_ptr, copy.get(), this->size_bytes() );
+      }
+      static constexpr bool is_legacy() { return LegacyAlign != 0; }
+
+      void* original_ptr;
+      std::unique_ptr<std::remove_cv_t<T>[]> copy = nullptr;
+      using dependent_type = T;
```

Comment:
I think this should also add `using proxy_type = span<T>;` to make it easier to determine what kind of `argument_proxy` it is elsewhere.